lib/perso.ml

```diff
@@ -3738,17 +4465,48 @@ let print_foreach conf base print_ast eval_expr =
                 List.iter
                   (fun ((p, _), baseprefix, can_merge) ->
                     if not can_merge then
-                      let env = ("#loop", Vint 0) :: env in
-                      let env = ("child_link", Vind p) :: env in
-                      let env = ("baseprefix", Vstring baseprefix) :: env in
-                      let env = ("p_link", Vbool true) :: env in
+                      let env =
+                        ("#loop", Vint 0) :: ("child_link", Vind p)
+                        :: ("baseprefix", Vstring baseprefix)
+                        :: ("p_link", Vbool true) :: env
+                      in
                       let ep = (p, true) in
                       List.iter (print_ast env ep) al)
                   children)
               (!GWPARAM_ITL.get_children' conf base (get_iper (fst ep)) fam isp)
         )
     | _ -> ()
   in
+  let print_foreach_descendant env al (p, _) count_paths =
+    let lev = match get_env "level" env with Vint lev -> lev | _ -> 0 in
+    let ip_l = match get_env "cousins" env with Vcousl l -> !l | _ -> [] in
```

Comment:
for both level and cousins, is the case `| _ ->` a failure?
If so please log it